### PR TITLE
Automatic late includes

### PR DESCRIPTION
**Pull request by jdemeyer**
This is a variation on #1650 which avoids new syntax. It is yet another attempt to solve the problem of accessing Cython variables from C, see [cython-users post](https://groups.google.com/forum/#!topic/cython-users/dA7ZtORhI9I).

With this pull request, includes can be "early" or "late". An early include is put in the Cython-generated `.c` file *before* Cython's declarations, a late include is put *after* Cython's declarations.

While #1650 proposed new syntax to classify an include as late, this pull request determines it automatically as follows: if a `cdef extern from "foo.h"` block is not empty and contains only variable or function declarations, then it is considered late. If it is empty, or it contains some kind of type declaration, then it is early.

CC @embray

### CODE REVIEW

File: Cython/Compiler/ModuleNode.py [hunk 1]
```diff
@@ -137,6 +138,7 @@ def process_implementation(self, options, result):
         env.return_type = PyrexTypes.c_void_type
         self.referenced_modules = []
         self.find_referenced_modules(env, self.referenced_modules, {})
+        env.fixup_includes()
```
Comment (robertwb) [inline Cython/Compiler/ModuleNode.py]: This seems a random place to stick fixup_includes. Maybe just do the deduplication in generate_includes?

**Comment by scoder** [comment]
Change looks ok. I agree with Robert that deduplication should just happen while late includes are being generated.

One drawback of this change is that it can be a bit surprising when arbitrarily unrelated things start misbehaving, e.g. after adding a new declaration. Since this feature is irrelevant for almost all Cython users, the underlying tradeoff stretches it a bit. Not all header files out there are 100% self-contained in their own includes, so changing their order is not risk-free. Can we hide this feature behind a directive, keeping it off by default?

**Comment by robertwb** [comment]
I'd rather reduce the number of directives we have. The only place I can see this going wrong is if someone has

```
cdef extern from "a.h":
    [non-type declarations]
cdef extern from "b.h":
    [type declarations]
```

but in fact `b.h` depends on `a.h`, which is possible but seems pretty rare (and even then, likely due to incomplete specification of what's actually in the headers). The workaround is to put an empty declaration to enforce the ordering (which is the justification for treating empty extern blocks as if they had type declarations). 

If we find we need to add a directive, let's at least keep the new behavior the default.

**Comment by scoder** [comment]
Ok

**Comment by jdemeyer** [comment]
So, you want me to change `fixup_includes` but you merged it anyway? Should I fix it in a follow-up? I also plan to add a testcase anyway.

Thanks for merging!

**Comment by scoder** [comment]
Robert already did both.

**Comment by jdemeyer** [comment]
I see that now. Great!

**Comment by cython-notifications** [comment]
FWIW, it exposed some oddities with functions that were defined but not
used outside of the late-included files. Not sure how best to resolve that.

**Comment by fried** [comment]
This has exposed some issues with Python.h being included first and defining some macros like T_BOOL that are later used as identifiers by C++ code. (ex: "apache::thrift::protocol::T_BOOL") 

Did we have Python.h included after extern statements in the Past?

**Comment by fried** [comment]
I have an extern statement like

`
cdef extern from 'some/module.h' namespace 'blah' nogil:
    void setSomeTimeout(int)
`

this include gets late bound, is this correct behavior

**Comment by fried** [comment]
This behavior just sounds like a bad idea, We can presume with a very strong assertion that this use case has never worked before.  So to opt existing code into this new behavior is a bad idea.  We should have some keyword to specially mark those include files that can be safely late included.

**Comment by jdemeyer** [comment]
> We should have some keyword to specially mark those include files that can be safely late included.

I agree but that idea was rejected by upstream Cython (see #1650).

**Comment by jdemeyer** [comment]
> This has exposed some issues with #include <structmember.h>
> being included first and defining some macros like T_BOOL that are later used as identifiers by C++ code. (ex: "apache::thrift::protocol::T_BOOL")

I don't see how that is related to this PR. If one header file defines a macro which is used somewhere else as identifier, you're in trouble anyway.

**Comment by fried** [comment]
But I can't control what cython includes as part of its code generation.
So its an issue with Cython.   And this Code did not break until after this change.

**Comment by yinghai** [comment]
You can also use this trick to avoid automatic late include
```
cdef extern from "spam.h":
           pass
```
This is actually documented in docs/src/userguide/external_C_code.rst.

**Comment by scoder** [comment]
> You can also use this trick ...

Yes, that's *exactly* what I dislike about this. We break user code, and then ask them to fix our breakage by adding obscure declarations to their code that no-one would ever understand without an additional comment saying "we had to do this because Cython ignores the order of our declarations here".

Given that the next release is already long overdue, I propose reverting this feature, rather than letting it spread out into the wild. We can then decide what to do about it. I created #2079 for this.

**Comment by jdemeyer** [comment]
> I propose reverting this feature

No no no no no no no no no no please! I understand that you want to fix things, but please don't break this feature! I was going to be so happy when Cython 0.28 came out because I could finally do what I have been trying for years.

**Comment by robertwb** [comment]
Could you explain further what the breakage was? I think understanding how
and why code is being broken, and how widespread it is, will determine the
path forward here (and long-term fix).

**Comment by fried** [comment]
The project was https://github.com/cupy/cupy used by https://github.com/chainer/chainer

on https://github.com/cupy/cupy/blob/master/cupy/cuda/driver.py
We now have a internal fork of where we add  
```
cdef extern from "cupy_cuda.h":
    pass
```
before the existing `cdef extern`

We do the same thing `cupy/cuda/cudnn.pyx`

I didn't work on it directly but I helped them get it building again, I looked up the internal post about the issue and he said "run cython compiler look at `driver.cpp` and see how `CUresult` is used in line 1129 while header of its definition is late included in line 1277"

**Pull request merged**